### PR TITLE
Triggered Camera

### PR DESCRIPTION
**Pull request by WilliamLewww**
# 🎉 Triggered Camera

Related Pull Requests:
* https://github.com/ignitionrobotics/ign-sensors/pull/194

Reference issue:
* https://github.com/ignitionrobotics/ign-sensors/issues/185

## Summary
Adds the SDF required to enable triggered cameras:
```
<camera>
  <triggered>true</triggered>
  <trigger_topic>/camera/trigger</trigger_topic>
  <horizontal_fov>1.05</horizontal_fov>
  <image>
    <width>256</width>
    <height>257</height>
  </image>
  <clip>
    <near>0.1</near>
    <far>10.0</far>
  </clip>
</camera>
```

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by ahcorde — DISMISSED** [review]
Missing tests

File: include/sdf/Camera.hh [hunk 1]
```diff
@@ -95,6 +95,10 @@ namespace sdf
     /// \param[in] _name Name of the sensor.
     public: void SetName(const std::string &_name);
 
+    /// \brief Get whether the camera is triggered by a topic.
+    /// \return If the camera is triggered by a topic.
```
Comment (nkoenig) [inline include/sdf/Camera.hh:99]:
```suggestion
    /// \return True if the camera is triggered by a topic.
```

File: src/Camera.cc [hunk 2]
```diff
@@ -55,6 +55,9 @@ class sdf::Camera::Implementation
   /// \brief Name of the camera.
   public: std::string name = "";
 
+  /// \brief True if the camera is triggered by a topic.
+  public: bool triggered;
```
Comment (nkoenig) [inline src/Camera.cc:59]:
```suggestion
  public: bool triggered{false};
```

File: src/Camera.cc [hunk 3]
```diff
@@ -402,6 +408,12 @@ void Camera::SetName(const std::string &_name)
   this->dataPtr->name = _name;
 }
 
+/////////////////////////////////////////////////
+bool Camera::Triggered() const
```
Comment (nkoenig) [inline src/Camera.cc:412]: Can you add a `void Camera::SetTriggered` function as well?

**Review by nkoenig — COMMENTED** [review]
Can you also add to the `Camera_TEST.cc` to cover these new features?

**Review by scpeters — COMMENTED** [review]
gazebo_ros_pkgs doesn't allow the trigger topic name to be configurable. I think we can do the same, unless someone thinks it is important to add this

**Comment by nkoenig** [comment]
> gazebo_ros_pkgs doesn't allow the trigger topic name to be configurable. I think we can do the same, unless someone thinks it is important to add this

I think you're referring to the `SetTrigger` function, which would be needed when creating the DOM programmatically.

**Comment by scpeters** [comment]
> > gazebo_ros_pkgs doesn't allow the trigger topic name to be configurable. I think we can do the same, unless someone thinks it is important to add this
> 
> I think you're referring to the `SetTrigger` function, which would be needed when creating the DOM programmatically.

I was referring to the fact that `gazebo_ros_pkgs` creates the trigger topic with a hard-coded `image_trigger` topic name:

* https://github.com/ros-simulation/gazebo_ros_pkgs/blob/ros2/gazebo_plugins/src/gazebo_ros_camera.cpp#L272
* https://github.com/ignitionrobotics/ign-sensors/pull/194/files#r803302851

My question is whether the trigger topic name should be configurable

**Comment by scpeters** [comment]
make a prerelease once this is merged to help test https://github.com/ignitionrobotics/ign-sensors/pull/194

**Comment by scpeters** [comment]
did we say the trigger topic name should be configurable?

**Review by scpeters — APPROVED** [review]

**Comment by WilliamLewww** [comment]
Re-requesting review to complete auto-merge.

**Comment by azeey** [comment]
I disabled the auto-merge. In general, we prefer to have a manual merge so that the author has a chance to modify the commit message such that it concisely captures the core ideas of the pull request.

From https://ignitionrobotics.org/docs/all/contributing#process
> Make sure the commit message concisely captures the core ideas of the pull request and contains all authors' signatures.

**Comment by chapulina** [comment]
> I disabled the auto-merge. In general, we prefer to have a manual merge so that the author has a chance to modify the commit message such that it concisely captures the core ideas of the pull request.

The auto-merge option allows editing the commit message, right? If the auto-merge is not working for us, we should disable it, but I think it should be compatible with our contribution process.

**Comment by azeey** [comment]
> The auto-merge option allows editing the commit message, right? If the auto-merge is not working for us, we should disable it, but I think it should be compatible with our contribution process.

Ah, it does. My mistake.

**Comment by scpeters** [comment]
I think @ahcorde concerns about the tests have been addressed, so I'll merge

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-04-13-fortress-edifice/1367/1

**Pull request merged**